rllib/core/sarl_trainer.py

```diff
@@ -0,0 +1,70 @@
+import abc
+
+
+class RLTrainer(abc.ABC):
+    # A base framework agnostic interface class for training RLModules
+    # module: RLModule: A reference to the RLModule that we are training
+
+    @abc.abstractmethod
+    def update(self, batch, **kwargs):
+        # Performs an update and returns anything that might be needed inside the
+        # algorithm.
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def make_module(self, module_config):
+        # a protected abstract method that creates the RLModule
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def make_optimizer(self, optimizer_config):
+        # a protected abstract method that creates the optimizer
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def compile_results(self, batch, fwd_out, update_out, **kwargs):
+        # a protected abstract method that computes the loss
+        raise NotImplementedError
+
+
+class SARLTrainer(RLTrainer, abc.ABC):
+    # A base framework agnostic class for training single-agent RLModules
+    # module: RLModule
+    def __init__(self, config):
+        # config: a dictionary of parameters
```

Comment:
Can we move self.config into the abstract base class' c'tor and have it assigned there?

Also, what's the type of `config`. We should start right away here with using `AlgorithmConfig` objects.